app/services/business_entity_service.rb

```diff
@@ -0,0 +1,56 @@
+class BusinessEntityService
+
+  def initialize(office, jurisdiction)
+    @office = office
+    @jurisdiction = jurisdiction
+    @timestamp = Time.zone.now
+    @business_entity = BusinessEntity.find_by(office: office,
+                                              jurisdiction: jurisdiction,
+                                              valid_to: nil)
+  end
+
+  def build(params)
+    build_business_entity(params)
+  end
+
+  def check_update(params)
+    build_business_entity(params)
+  end
+
+  def persist_update!(business_entity)
+    return false if business_entity.nil? || @business_entity.nil?
+    if duplicate_needed?(business_entity)
+      @business_entity.assign_attributes(valid_to: @timestamp)
+      save_entities_in_transaction(business_entity)
+    else
+      @business_entity.assign_attributes(name: business_entity.name)
+      @business_entity.save
+    end
+  end
+
+  private
+
+  def build_business_entity(params)
+    BusinessEntity.new(office: @office,
+                       jurisdiction: @jurisdiction,
+                       name: params[:name],
+                       code: params[:code],
+                       valid_from: @timestamp)
+  end
+
+  def duplicate_needed?(business_entity)
+    business_entity != @business_entity &&
+      (both_entities_present?(business_entity) && business_entity.code != @business_entity.code)
+  end
```

Comment:
This is a bit confusing as it doesn't read easily. I'd break it down into smaller chunks.

Maybe something like this:

```ruby
def duplicate_needed?(business_entity)
  entities_match?(business_entity) &&
    both_entities_present_and_codes_match?(business_entity)
end

def entities_match?(business_entity)
  business_entity != @business_entity
end

def codes_match?(business_entity)
  business_entity.code != @business_entity.code
end

def both_entities_present_and_codes_match?(business_entity)
  both_entities_present?(business_entity) && codes_match?(business_entity)
end
```